### PR TITLE
(WIP) Prototype of offline inline caching

### PR DESCRIPTION
**Pull request by Duhemm**
This PR adds a prototype of offline inline caching, which works in two steps:
### 1. The application is compiled in profiling mode

First, we collect information about the runtime type of the receivers of virtual method calls at runtime.  This way, we can determine what are the most frequent types that we encounter for each virtual call site.

To enable profiling, set `nativeProfileDispatch := true` in sbt, and optionally set `nativeProfileInfo := Some(file(...))` to have the profiling information written on disk.
### 2. The application is recompiled, leveraging the information we collected

If the compiler finds the file specified in `nativeProfileInfo`, then it will use the information it contains to rewrite call sites to avoid virtual dispatch.
#### Example

Given this information collected at runtime (the first number is the type id, the second is the number of occurrences):

```
= src.2:addToGraph_unit:
    3 (100000)
    4 (100000)
    6 (5)
    7 (100001)
```

This definition, which contains a call to `addToGraph_unit`:

```
def @Constraint::addConstraint_unit : (class @Constraint) => unit {
  %src.1(%src.0: class @Constraint):
    %src.2 = method[(class @Constraint) => unit] %src.0: class @Constraint, @Constraint::addToGraph_unit
    %src.3 = call[(class @Constraint) => unit] %src.2: ptr(%src.0: class @Constraint)
    ...
}
```

becomes

```
def @Constraint::addConstraint_unit : (class @Constraint) => unit {
  %src.1(%src.0: class @Constraint):
    %tx.2894 = load[ptr] %src.0: class @Constraint
    jump %tx.2905
  %tx.2905:
    %tx.2899 = ieq[ptr] %tx.2894: ptr, @type.class.StayConstraint
    if %tx.2899: bool then %tx.2895 else %tx.2904
  %tx.2904:
    %tx.2900 = ieq[ptr] %tx.2894: ptr, @type.class.ScaleConstraint
    if %tx.2900: bool then %tx.2897 else %tx.2903
  %tx.2903:
    %tx.2901 = method[(class @Constraint) => unit] %src.0: class @Constraint, @Constraint::addToGraph_unit
    %tx.2902 = call[(class @Constraint) => unit] %tx.2901: ptr(%src.0: class @Constraint)
    jump %tx.2893(%tx.2902: unit)
  %tx.2895:
    %tx.2896 = call[(class @Constraint) => unit] @UnaryConstraint::addToGraph_unit(%src.0: class @Constraint)
    jump %tx.2893(%tx.2896: unit)
  %tx.2897:
    %tx.2898 = call[(class @Constraint) => unit] @ScaleConstraint::addToGraph_unit(%src.0: class @Constraint)
    jump %tx.2893(%tx.2898: unit)
  %tx.2903:
    %tx.2901 = method[(class @Constraint) => unit] %src.0: class @Constraint, @Constraint::addToGraph_unit
    %tx.2902 = call[(class @Constraint) => unit] %tx.2901: ptr(%src.0: class @Constraint)
    jump %tx.2893(%tx.2902: unit)
  %tx.2893(%src.3: unit):
    ...
}
```
### 3. Performance improvements

The performance improvement is not very impressive on the `deltablue` ([here](https://github.com/Duhemm/Deltablue)) benchmark, though this optimization reduces the number of virtual calls from 20 millions to 800 thousands.
#### Without inline caching, compiled with `-O2`:

```
❯ tim ./vanilla-o2-out                                                                                                                                          ./vanilla-o2-out  0.94s user 0.08s system 118% cpu 0.858 total
./vanilla-o2-out  0.92s user 0.07s system 120% cpu 0.820 total
./vanilla-o2-out  0.93s user 0.08s system 118% cpu 0.854 total
./vanilla-o2-out  0.93s user 0.07s system 118% cpu 0.846 total
./vanilla-o2-out  0.93s user 0.08s system 119% cpu 0.839 total
./vanilla-o2-out  0.94s user 0.07s system 119% cpu 0.842 total
./vanilla-o2-out  0.96s user 0.08s system 118% cpu 0.872 total
./vanilla-o2-out  0.96s user 0.07s system 118% cpu 0.869 total
./vanilla-o2-out  0.93s user 0.07s system 119% cpu 0.839 total
./vanilla-o2-out  0.92s user 0.07s system 119% cpu 0.826 total
```
#### With inline caching, compiled with `-O2`:

```
❯ tim ./opt-o2-out                                                                                                                                              ./opt-o2-out  0.91s user 0.07s system 119% cpu 0.818 total
./opt-o2-out  0.92s user 0.07s system 119% cpu 0.828 total
./opt-o2-out  0.90s user 0.07s system 119% cpu 0.813 total
./opt-o2-out  0.91s user 0.07s system 119% cpu 0.824 total
./opt-o2-out  0.92s user 0.07s system 118% cpu 0.838 total
./opt-o2-out  0.93s user 0.07s system 120% cpu 0.834 total
./opt-o2-out  0.94s user 0.08s system 118% cpu 0.868 total
./opt-o2-out  0.90s user 0.07s system 119% cpu 0.808 total
./opt-o2-out  0.92s user 0.08s system 118% cpu 0.844 total
./opt-o2-out  0.90s user 0.07s system 117% cpu 0.820 total
```
### ~~3. Problems~~

~~I'm not seeing any performance improvements. We saw performance improvements thanks to LLVM's optimizer (we compared a non-optimized version without inline caching against an optimized (-O2) version with inline caching). The number of virtual calls is reduced on my benchmark from 20 millions to 800 thousands when inlining the 2 most frequent candidates.~~

~~From the NIR, is there any obvious reason why we're not seeing any performance improvement (nor degradation)?~~

### CODE REVIEW

**Comment by densh** [comment]
Can you also include the code of the benchmark you used here?

File: tools/src/main/scala/scala/scalanative/compiler/tools/DispatchInfoParser.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package scala.scalanative
+package compiler
+package tools
+
+import scala.util.parsing.combinator._
+import scala.util.matching.Regex
+
+object DispatchInfoParser extends JavaTokenParsers {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/tools/DispatchInfoParser.scala]: Scala's parser combinators are not being actively maintained. Using `fastparse` for this kind of things is probably a better idea.

**Comment by pbatko** [comment]
Is the idea about having more code and branching but less memory accesses (to dispatch table)?

I don't know how efficient is normal (without this optimization) dynamic dispatch in scala native (two loads? load dispatch table address, load function pointer), but maybe the optimization saves one load there (in the common case) but at the same time one extra load is needed to check the type of the receiver?

**Comment by densh** [comment]
Needs a rebase due to scalafmt upgrade.

File: tools/src/main/scala/scala/scalanative/compiler/tools/DispatchInfoParser.scala (hunk 1 above)
Comment (Duhemm) [inline tools/src/main/scala/scala/scalanative/compiler/tools/DispatchInfoParser.scala]: I migrated to fastparse

**Comment by Duhemm** [comment]
I updated the PR description to reflect our last discussion.

**Comment by densh** [comment]
Status here: me and @Duhemm are assembling benchmarking suite based on https://github.com/smarr/are-we-fast-yet and https://github.com/sjrd/scalajs-benchmarks to properly evaluate performance implications of this PR.

**Comment by densh** [comment]
Now that #405 is merged, this needs a rebase.

**Comment by Duhemm** [comment]
Just finished rebasing.

**Comment by densh** [comment]
Given that this experiment is still ongoing, lets postpone the actual PR for now.

**Pull request closed without merging**